core/charm/baseselector.go

```diff
@@ -0,0 +1,194 @@
+		logger:              cfg.Logger,
+		usingImageID:        cfg.UsingImageID,
+	}
+	bs.supportedBases, err = bs.validate(cfg.SupportedCharmBases, workloadBases)
+	if err != nil {
+		return BaseSelector{}, errors.Trace(err)
+	}
+	return bs, nil
+}
+
+// TODO(nvinuesa): The force flag is only valid if the requestedBase is specified
+// or to force the deploy of a LXD profile that doesn't pass validation, this
+// should be added to these validation checks.
+func (s BaseSelector) validate(supportedCharmBases, supportedJujuBases []series.Base) ([]series.Base, error) {
+	// If the image-id constraint is provided then base must be explicitly
+	// provided either by flag either by model-config default base.
+	if s.logger == nil {
+		return nil, errors.NotValidf("empty Logger")
+	}
+	if s.usingImageID && s.requestedBase.Empty() && !s.explicitDefaultBase {
+		return nil, errors.Forbiddenf("base must be explicitly provided when image-id constraint is used")
+	}
+	if len(supportedCharmBases) == 0 {
+		return nil, errors.NotValidf("charm does not define any bases,")
+	}
+	if len(supportedJujuBases) == 0 {
+		return nil, errors.BadRequestf("programming error: no juju supported bases")
+	}
+	// Verify that the charm supported bases include at least one juju
+	// supported base.
+	var supportedBases []series.Base
+	for _, charmBase := range supportedCharmBases {
+		for _, jujuCharmBase := range supportedJujuBases {
+			if jujuCharmBase.IsCompatible(charmBase) {
+				supportedBases = append(supportedBases, charmBase)
+				s.logger.Infof(msgUserRequestedBase, charmBase)
+			}
+		}
+	}
+	if len(supportedBases) == 0 {
+		return nil, errors.NotSupportedf("the charm defined bases %q", printBases(supportedCharmBases))
+	}
+	return supportedBases, nil
+}
+
+// CharmBase determines what base to use with a charm.
+// Order of preference is:
+//   - user requested with --base or defined by bundle when deploying
+//   - model default, if set, acts like --base
+//   - juju default ubuntu LTS from charm manifest
+//   - first base listed in the charm manifest
+//   - in the case of local charms with no manifest nor base in metadata,
+//     base must be provided by the user.
+func (s BaseSelector) CharmBase() (selectedBase series.Base, err error) {
+	// TODO(sidecar): handle systems
+
+	// TODO (hml) 2023-05-16
+	// BaseSelector needs refinement. It is currently a copy of
+	// SeriesSelector, however it does too much for too many
+	// cases.
+
+	// User has requested a base with --base.
+	if !s.requestedBase.Empty() {
+		return s.userRequested(s.requestedBase)
+	}
+
+	// Use model default base, if explicitly set and supported by the charm.
+	// Cannot guarantee that the requestedBase is either a user supplied base or
+	// the DefaultBase model config if supplied.
+	if s.explicitDefaultBase {
+		return s.userRequested(s.defaultBase)
+	}
+
+	// Prefer latest Ubuntu LTS.
+	preferredBase, err := BaseForCharm(series.LatestLTSBase(), s.supportedBases)
```

Comment:
If the charm doesn't provide any `supportedBases`, then anything after this point is dead, right? because `series.LatestLTSBase()` will always return a valid base.

Also, if we don't know what the charm supports, then we should probably require `--force` like the series selector.